### PR TITLE
feat(autoscaling): Auto Scaling Group with Launch Template

### LINKED ISSUE #6734: Add support for Launch Templates
Currently the CDK configures EC2 autoscaling using  LaunchConfigurations, we should add support for  [Launch Templates](https://docs.aws.amazon.com/autoscaling/ec2/userguide/LaunchTemplates.html)  which enables more advanced features such as: 

* [Mixed auto scaling groups](https://docs.aws.amazon.com/autoscaling/ec2/userguide/asg-purchase-options.html)
* Versioning

### PR DESCRIPTION
**Pull request by Martin1994**
Add support to launch an Auto Scaling Group from a Launch Template rather than a launch configuration.

Closes #6734.

----

High level designs:

The current AutoScalingGroup L2 construct, unfortunately, is deeply coupled with LaunchConfiguration. Therefore adding LaunchTemplate to the existing construct is not trivial.

There are two different approaches to support LaunchTemplate:
1. Implement another brand new L2 construct with `IAutoScalingGroup` interface. The shared logic between the old and the new one such as adding scaling policies can be extracted to a common parent class.
2. Add LaunchTemplate related interface to the existing L2 construct with minimum breaking changes.

Adding a new construct is always guaranteed to be a clean solution code-wise, but it can also cause confusion from the end user about scattered CDK implementation on the same CFN resource, and on the other hand breaks the existing libraries consuming the existing `AutoScalingGroup` construct rather than the `IAutoScalingGroup` interface.

Adding LT to the existing construct, as what this PR does, however may change the API behaviour when LT is used. For example, the implementation in this PR turns `AutoScalingGroup.connections` from a public field to a public getter, and will throw an error when the ASG is created from a LT reference, which means existing libraries taking an `AutoScalingGroup` instance as an `IConnectable` will get runtime error in this case. Existing code (i.e. ASG created from the L2 construct with a LC rather than a LT) won't break with this change.

This PR picks the latter approach since I believe it provides a better experience overall, mainly because of its continuity.

----

BREAKING CHANGE: Properties relying on a launch configuration are now either optional or turned into throwable getters
  * **autoscaling:** `AutoScalingGroupProps.instanceType` is now optional
  * **autoscaling:** `AutoScalingGroupProps.machineImage` is now optional
  * **autoscaling:** `AutoScalingGroup.connections` is now a throwable getter
  * **autoscaling:** `AutoScalingGroup.role` is now a throwable getter
  * **autoscaling:** `AutoScalingGroup.userData` is now a throwable getter

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by fiftysquid** [comment]
Actually being able to use Launch Templates would be neat.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (jungseoklee) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:337]: I am wondering if it is better to have a separate file for mixed instance policy because this file is getting bigger.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (jungseoklee) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1629]: I am wondering if it is better to have a separate file for launch template and launch configuration because this file is getting bigger.

**Comment by jungseoklee** [comment]
The approach looks reasonable. As mentioned, this won't break existing LC users, which is what I like.

**Comment by jungseoklee** [comment]
The approach looks reasonable. As mentioned, this won't break existing LC users, which is what I like.

**Comment by jungseoklee** [comment]
Can we get attention from maintainers? I see more than 100 thumbs up for this feature, so it would be great if we can start reviewing this PR.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:485]: It becomes optional after the introduction of `InstanceRequirements` in Oct 2021. Shall we include attribute-based instance type selection in this PR or have a follow-up one for that?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1629]: I am not sure how practical it is since this design puts LC and LT under one single class, and I don't think we can break a class definition into multiple files like `partial class` in C#.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (jungseoklee) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:485]: IMHO, follow-up should be good because it is more important to unblock using Launch Template.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1478]: Can we improve this comment to explain what the `UserData` does and how it is used?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1493]: Can we update this comment to explain what this `Role` is used for? eg if it's an execution role, that should be mentioned here (also, this variable probably needs a better name than `role`; perhaps `executionRole` or `instanceExecutionRole`, if that's what it represents).

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1463]: Can we improve this comment to explain how this property it used, or remove it entirely?

**Review by comcalvi — CHANGES_REQUESTED** [review]
Good first attempt! We're a few iterations from merging this in, but this is looking great so far.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (jungseoklee) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:418]: @comcalvi, do we need to drop dash, `-`, when commenting the default value, which is what I got from @kaizen3031593 during code review on [this PR](https://github.com/aws/aws-cdk/pull/19214/files)?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1226]: This logic is not introduced by this PR. From what I understand, `addSecurityGroup` pushes new security groups into this list. `CfnLaunchConfiguration` takes in a list of SGs.

This branch is not used if a Launch Template is passed in.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1164]: Line 1246 is in a if branch when no launch template is provided. A launch config is mutually exclusive with a launch template.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1493]: I cannot change the getter name here (same as `userData`) because it was a class property. `role` and `userData` are turned into getters so that they can throw exceptions when a Launch Template reference is given.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1478]: I can borrow some words from the public doc. However more importantly I think it's better to document the fact that this getter can throw error. Same as the `role` getter.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1463]: These getters were class properties. I copy-pasted their doc as is. To keep the backward compatibility I cannot really remove or rename them.

I can have a try improving the doc here in the next revision.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:418]: I will follow that comment too

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:478]: This is related to the `InstanceType` comment.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1195]: I won't move the validations in the other branch for launch config btw, because they are short, and some subsequent code requires the non-null check there.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1463]: Also this implements `IConnections`. It is a common practice to document something like _This property implements `IConnections`_?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:418]: From what I understand, if the default value is some specific value, it should be used without a dash. If `@default` only follows a pure documentation then a dash is needed otherwise the first word will be recognized as the value.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: If the addition of the property is intended to implement `IConnections`, then you can add `implements IConnections` to the class declaration.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: yes, the `-` should be dropped

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: It had been there before this change.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1688]: Good catch.. That was a missing branch.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1194]: This was from the existing code. 

https://github.com/aws/aws-cdk/blob/77a5fa1f0ecc2dcaa270212c8d61479f9a1c8ccd/packages/%40aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts#L967

`props.allowAllOutbound` is `boolean?` so I think `!!props.allowAllOutbound` would be more clear

Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1194]: nvm its default value is `true` so `!== false` is correct. I will leave it untouched.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Ah. No need to add a comment stating that the property implements `IConnections` then.

**Comment by Martin1994** [comment]
Any update?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1178]: this code:

```
      vpc: new ec2.Vpc(this, 'template-vpc'),
      mixedInstancesPolicy: {
        launchTemplate: new ec2.LaunchTemplate(this, 'template-mixed', {
          instanceType: new ec2.InstanceType('a1.2xlarge')
        })
      }
    });
```

Results in a deploy time error:

```
You must use a valid fully-formed launch template. The request must contain the parameter ImageId (Service: AmazonAutoScaling; Status Code: 400; Error Code: ValidationError;
```

This error only happens if the launch template is specified here as part of an autoscaling group, and does not happen when a launch template is specified on its own; thus, we should ensure the user has passed an `imageId` here, and throw if they have not.

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1542]: need a test case verifying that it can't be created without `imageId` as well.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1184]: This `if else` statement (the `if` on line 1165, and the `else` on 1184) is far too big. Can we extract the code in each case to its own function? So we'd have:

```
    if (props.launchTemplate || props.mixedInstancesPolicy) {
      handleIfCase();
    } else {
      handleElseCase();
    }
```

(of course, with better-named functions)

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1608]: I like that we have these test cases like this, but can we add some slight validation to the error thrown, to ensure it at least partially matches the error message?

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1612]: can we validate that the error thrown here at least partially matches the error message?

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1616]: can we validate that the error thrown here at least partially matches the error message?

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1620]: can we validate that the error thrown here at least partially matches the error message?

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1586]: can we validate that the error thrown here at least partially matches the error message?

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1574]: can we validate that the error thrown here at least partially matches the error message? Same comment on all of these `.toThrow()` calls.

**Review by comcalvi — CHANGES_REQUESTED** [review]
This is looking better! In general, when a PR is ready for review, please re-request the review on github

File: packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts
Comment (Martin1994) [reply packages/@aws-cdk/aws-autoscaling/test/auto-scaling-group.test.ts:1542]: `ImageId` is inferred from `MachineImage` so the previous test case already covers this.

**Review by comcalvi — DISMISSED** [review]
This is pretty much perfect, just one small comment.

**Comment by robertd** [comment]
Thanks for all the hard work you put in on this PR @Martin1994 🙌

**Comment by Martin1994** [comment]
Just to confirm, GitHub can rebase, squash and merge this PR right?

**Comment by robertd** [comment]
@Martin1994 Once @comcalvi approves this PR GitHub Actions will do all the grunt work.

**Review by comcalvi — DISMISSED** [review]
Awesome work @Martin1994!

**Comment by robertd** [comment]
Very excited for this PR!!! Thanks everyone. 🙏

**Comment by jungseoklee** [comment]
Very excited. Great thanks, everyone!

**Comment by robertd** [comment]
Thanks for stepping in @comcalvi 🙌

**Review by comcalvi — APPROVED** [review]

**Pull request merged**
